[PyTorch Mobile] Record dtypes for tensors used in kernel function implementations

dhruvbird opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48826 [PyTorch Mobile] Record dtypes for tensors used in kernel function implementations**

This change updates various macros to pass in the kernel tag string (`const char*`) into the macro that sets up the `case` statement for the dtype switch. This macro already receives the dtype (enum) which we also need.

There are 2 phases we need to build out for the ftype tracing to work:
1. Recording Phase
2. Conditional Compilation Phase

For this most part, this change is trying to focus on [1] (The Recording Phase) and sets up a new `RecordScope` enum value to track kernel dtypes. This code is compiled in only if a specific macro is defined (since this is an **extremely** hot code path, and even the slightest regression here can cause tremendous slow down overall).

I have only added a skeleton of the phase [2] (Conditional Compilation Phase) and there is a no-op `constexpr` method that selects every dtype in the kernel implementation. In subsequent diffs, this will be updated to point to a code-generated function based on the result of tracing the models that were requested.

Differential Revision: [D24220926](https://our.internmc.facebook.com/intern/diff/D24220926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24220926/)!

dhruvbird committed:
[PyTorch Mobile] Record dtypes for tensors used in kernel function implementations

This change updates various macros to pass in the kernel tag string (`const char*`) into the macro that sets up the `case` statement for the dtype switch. This macro already receives the dtype (enum) which we also need.

There are 2 phases we need to build out for the ftype tracing to work:
1. Recording Phase
2. Conditional Compilation Phase

For this most part, this change is trying to focus on [1] (The Recording Phase) and sets up a new `RecordScope` enum value to track kernel dtypes. This code is compiled in only if a specific macro is defined (since this is an **extremely** hot code path, and even the slightest regression here can cause tremendous slow down overall).

I have only added a skeleton of the phase [2] (Conditional Compilation Phase) and there is a no-op `constexpr` method that selects every dtype in the kernel implementation. In subsequent diffs, this will be updated to point to a code-generated function based on the result of tracing the models that were requested.

Differential Revision: [D24220926](https://our.internmc.facebook.com/intern/diff/D24220926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24220926/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch Mobile] Record dtypes for tensors used in kernel function implementations"


This change updates various macros to pass in the kernel tag string (`const char*`) into the macro that sets up the `case` statement for the dtype switch. This macro already receives the dtype (enum) which we also need.

There are 2 phases we need to build out for the ftype tracing to work:
1. Recording Phase
2. Conditional Compilation Phase

For this most part, this change is trying to focus on [1] (The Recording Phase) and sets up a new `RecordScope` enum value to track kernel dtypes. This code is compiled in only if a specific macro is defined (since this is an **extremely** hot code path, and even the slightest regression here can cause tremendous slow down overall).

I have only added a skeleton of the phase [2] (Conditional Compilation Phase) and there is a no-op `constexpr` method that selects every dtype in the kernel implementation. In subsequent diffs, this will be updated to point to a code-generated function based on the result of tracing the models that were requested.

Differential Revision: [D24220926](https://our.internmc.facebook.com/intern/diff/D24220926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24220926/)!

[ghstack-poisoned]

dhruvbird commented on aten/src/ATen/Dispatch.h:
```diff
@@ -2,17 +2,55 @@
+#include <c10/util/Metaprogramming.h>
 #include <c10/util/complex.h>
+#include <c10/util/string_view.h>
+#include <iostream>
```
oops. Will remove.

ilia-cher commented:
cc. @ezyang @ngimel

ezyang commented:
> This code is compiled in only if a specific macro is defined (since this is an extremely hot code path, and even the slightest regression here can cause tremendous slow down overall).

So what, does this mean you have to build PyTorch under a custom build config to be able to trace? Sounds like a pain...

ezyang commented:
So... what happens if someone accidentally uses the same NAME for two distinct macro invocations? I guess it will still do a correct conservative approximation, because if one of them pings during tracing, you'll record it, and then ALL of the name sites will get flipped on. Still, seems like a kind of very annoying imprecision to debug if someone messes this up, since magically some unrelated fragment will still get included in the binary and you just *won't know about it.*

ezyang reviewed (approved):
This looks basically reasonable but there should be a Note somewhere explaining how the apparatus works, and instructing people about why they should take care specifying NAME in the dispatch macros.

dhruvbird commented:
> > This code is compiled in only if a specific macro is defined (since this is an extremely hot code path, and even the slightest regression here can cause tremendous slow down overall).
> 
> So what, does this mean you have to build PyTorch under a custom build config to be able to trace? Sounds like a pain...
> 
> EDIT: Based on comments in Phabricator, this seems to be exactly the plan. But since this is mobile land we already have a bajillion build configurations, what's one more hehe

Yep, it's definitely a special case, but all of this is hidden from the users by scripting and automation, so for the most part no engineers would really need to know about this except for someone on the team.

I have additionally also documented in the rev. that I shall just put up.

Thanks for calling this out (and also the other comments which have been addressed).

dhruvbird committed:
Update on "[PyTorch Mobile] Record dtypes for tensors used in kernel function implementations"


This change updates various macros to pass in the kernel tag string (`const char*`) into the macro that sets up the `case` statement for the dtype switch. This macro already receives the dtype (enum) which we also need.

There are 2 phases we need to build out for the ftype tracing to work:
1. Recording Phase
2. Conditional Compilation Phase

For this most part, this change is trying to focus on [1] (The Recording Phase) and sets up a new `RecordScope` enum value to track kernel dtypes. This code is compiled in only if a specific macro is defined (since this is an **extremely** hot code path, and even the slightest regression here can cause tremendous slow down overall).

I have only added a skeleton of the phase [2] (Conditional Compilation Phase) and there is a no-op `constexpr` method that selects every dtype in the kernel implementation. In subsequent diffs, this will be updated to point to a code-generated function based on the result of tracing the models that were requested.

Differential Revision: [D24220926](https://our.internmc.facebook.com/intern/diff/D24220926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24220926/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch Mobile] Record dtypes for tensors used in kernel function implementations"


This change updates various macros to pass in the kernel tag string (`const char*`) into the macro that sets up the `case` statement for the dtype switch. This macro already receives the dtype (enum) which we also need.

There are 2 phases we need to build out for the ftype tracing to work:
1. Recording Phase
2. Conditional Compilation Phase

For this most part, this change is trying to focus on [1] (The Recording Phase) and sets up a new `RecordScope` enum value to track kernel dtypes. This code is compiled in only if a specific macro is defined (since this is an **extremely** hot code path, and even the slightest regression here can cause tremendous slow down overall).

I have only added a skeleton of the phase [2] (Conditional Compilation Phase) and there is a no-op `constexpr` method that selects every dtype in the kernel implementation. In subsequent diffs, this will be updated to point to a code-generated function based on the result of tracing the models that were requested.

Differential Revision: [D24220926](https://our.internmc.facebook.com/intern/diff/D24220926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24220926/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch Mobile] Record dtypes for tensors used in kernel function implementations"


This change updates various macros to pass in the kernel tag string (`const char*`) into the macro that sets up the `case` statement for the dtype switch. This macro already receives the dtype (enum) which we also need.

There are 2 phases we need to build out for the ftype tracing to work:
1. Recording Phase
2. Conditional Compilation Phase

For this most part, this change is trying to focus on [1] (The Recording Phase) and sets up a new `RecordScope` enum value to track kernel dtypes. This code is compiled in only if a specific macro is defined (since this is an **extremely** hot code path, and even the slightest regression here can cause tremendous slow down overall).

I have only added a skeleton of the phase [2] (Conditional Compilation Phase) and there is a no-op `constexpr` method that selects every dtype in the kernel implementation. In subsequent diffs, this will be updated to point to a code-generated function based on the result of tracing the models that were requested.

Differential Revision: [D24220926](https://our.internmc.facebook.com/intern/diff/D24220926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24220926/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch Mobile] Record dtypes for tensors used in kernel function implementations"


This change updates various macros to pass in the kernel tag string (`const char*`) into the macro that sets up the `case` statement for the dtype switch. This macro already receives the dtype (enum) which we also need.

There are 2 phases we need to build out for the ftype tracing to work:
1. Recording Phase
2. Conditional Compilation Phase

For this most part, this change is trying to focus on [1] (The Recording Phase) and sets up a new `RecordScope` enum value to track kernel dtypes. This code is compiled in only if a specific macro is defined (since this is an **extremely** hot code path, and even the slightest regression here can cause tremendous slow down overall).

I have only added a skeleton of the phase [2] (Conditional Compilation Phase) and there is a no-op `constexpr` method that selects every dtype in the kernel implementation. In subsequent diffs, this will be updated to point to a code-generated function based on the result of tracing the models that were requested.

Differential Revision: [D24220926](https://our.internmc.facebook.com/intern/diff/D24220926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24220926/)!

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/Dispatch.h:
```diff
@@ -2,17 +2,59 @@
+ * included based on build time constants generated from tracing model
+ * execution. This method will be implmeneted via code-generation and
+ * included in this file when code-gen is ready.
+ */
```
Hey @dhruvbird, is this comment still relevant? I see in master now we still have the comment, but we also have a mobile ifdef that adds the codegen'd version of the function:

```
#ifdef TEMPLATE_SELECTIVE_BUILD
#include <ATen/selected_mobile_ops.h>
#else
namespace at {
/**
 * The method should_include_kernel_dtype() returns true/false
 * based on whether the switching code for a specific dtype should be
 * included based on build time constants generated from tracing model
 * execution. This method will be implmeneted via code-generation and
 * included in this file when code-gen is ready.
 */
inline constexpr bool should_include_kernel_dtype(
  const char *kernel_tag_str,
  at::ScalarType scalar_type
) {
  return true;
}
}
#endif
```

So is `should_include_kernel_dtype()` always going to be a no-op for server builds? (cc @guoyejun)

dhruvbird replied:
Yes, the comment is slightly out of date. Basically this code:

```
inline constexpr bool should_include_kernel_dtype(
  const char *kernel_tag_str,
  at::ScalarType scalar_type
) {
  return true;
}
```

will be used for server builds, and the code from the included file `<ATen/selected_mobile_ops.h>` will be used for mobile builds.

> So is should_include_kernel_dtype() always going to be a no-op for server builds? (cc @guoyejun)

Yes - basically any build that doesn't define the macro `TEMPLATE_SELECTIVE_BUILD`.

bdhirsh replied:
got it thanks :)

guoyejun replied:
thanks
